monitoring/ceph-mixin/prometheus_alerts.libsonnet

```diff
@@ -0,0 +1,695 @@
+(import 'config.libsonnet')
+{
+  MultiClusterQuery()::
+    if $._config.showMultiCluster
+    then 'cluster,'
+    else '',
+
+  MultiClusterSummary()::
+    if $._config.showMultiCluster
+    then ' on cluster {{ $labels.cluster }}'
+    else '',
+
+  groups+: [
+    {
+      name: 'cluster health',
+      rules: [
+        {
+          alert: 'CephHealthError',
+          'for': '5m',
+          expr: 'ceph_health_status == 2',
+          labels: { severity: 'critical', type: 'ceph_default', oid: '1.3.6.1.4.1.50495.1.2.1.2.1' },
+          annotations: {
+            summary: 'Ceph is in an ERROR state%(cluster)s' % $.MultiClusterSummary(),
```

Comment:
What version of jsonnet do you need to work with the `%` operator? I tried with mine and it didn't work.

Reply:
I use v0.18.0, weird that it did not work for you how did you call jsonnet? To enable multicluster change `showMultiCluster` to `true` in `ceph-mixin/config.libsonnnet`, I will add clarify this in the README

Reply:
Other mixins (Node, Mimir, Kubernetes) take a similar config based approach and from what I can tell there is not really a good method to dynamically extend queries. However, if the cluster label would be defined in every Ceph deployment it would probably be possible have the same alert file for multi and single cluster. Except the cluster label is not defined with the default cephadm install for example so it would break for many users. Also the building and deployment of alerts can be automated in pipelines so it should not be a big problem

Reply:
> I use v0.18.0, weird that it did not work for you how did you call jsonnet?

It worked now. I had some formatting issue. Thanks!

> To enable multicluster change showMultiCluster to true in ceph-mixin/config.libsonnnet, I will add clarify this in the README

I'm not familiar with other mixins, but for building purposes I find that having to update a config file is less practical than providing command-line flags.

What about using [jsonnet top-level args](https://jsonnet.org/learning/tutorial.html#parameterize-entire-config)? I think that fits your use-case here and it allows for CLI settings.

> However, if the cluster label would be defined in every Ceph deployment it would probably be possible have the same alert file for multi and single cluster. Except the cluster label is not defined with the default cephadm install for example so it would break for many users. Also the building and deployment of alerts can be automated in pipelines so it should not be a big problem

I assume that multi-cluster users won't use a single Prometheus instance and will be have some kind of federation (via remote writes or scaping). In that case the problem could be solved [with a new `cluster` label in a couple of ways via Prometheus config](https://stackoverflow.com/questions/47628532/how-can-we-add-extra-label-to-prometheus-metrics).

However if we want to keep the queries working for single-cluster deployments, we should still add a `cluster` label. I'd avoid generating the label in the metrics exporter dump itself, since it'll just be static data repeated over and over, and I'd add the label in the local Prometheus config (as mentioned above) or via [Service Discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config), especially given that [Cephadm is moving to SD](https://github.com/ceph/ceph/pull/46400/) for Node-exporter, HAProxy and mgr/Prometheus metrics.

An alternative to adding a new `cluster` label would be to update one of the existing labels to convey the cluster fsid (what about the `job=node.<fsid>`.

BTW and a bit off-topic, cluster fsid's are not the friendlies way to identify multiple clusters. We (Ceph Dashboard) team had a recent discussion about auto-generating/allowing users to customize friendly names for clusters. The fsid would continue being the source of truth, but perhaps we could expose this friendly cluster ID instead (as [Docker does](https://github.com/moby/moby/blob/master/pkg/namesgenerator/names-generator.go)): `optimistic-cuttlefish` or `condescending-kraken`.

Reply:
> I'm not familiar with other mixins, but for building purposes I find that having to update a config file is less practical than providing command-line flags.
> 
> What about using [jsonnet top-level args](https://jsonnet.org/learning/tutorial.html#parameterize-entire-config)? I think that fits your use-case here and it allows for CLI settings.

If ceph-mixin starts to use tlsa instead of the standard config object, it means that we won't be able to easily configure the mixin in automation via Jsonnet anymore (i.e.: https://github.com/cernops/ceph-monit/blob/main/mixins/ceph-mixin.libsonnet#L2-L4) and that's for me one of the core value of jsonnet/mixin to embrace automation and re-usability. And well even for people that don't do this and just want to tweak a setting or two in a onetime command I don't think that editing this config file and running is that a hassle IMO.

> I assume that multi-cluster users won't use a single Prometheus instance and will be have some kind of federation (via remote writes or scaping). In that case the problem could be solved [with a new `cluster` label in a couple of ways via Prometheus config](https://stackoverflow.com/questions/47628532/how-can-we-add-extra-label-to-prometheus-metrics).

Well CERN does. We have a single Prometheus instance collecting 15-20 Ceph cluster with all the associated node-exporter and a bit more. We might split it up a bit in the future (and probably also introduce HA pairs) but one Prometheus would probably continue scrapping 5-10 clusters or so.

As you said though there are many ways to add this label in Prometheus depending on your setup indeed.

> However if we want to keep the queries working for single-cluster deployments, we should still add a `cluster` label. I'd avoid generating the label in the metrics exporter dump itself, since it'll just be static data repeated over and over, and I'd add the label in the local Prometheus config (as mentioned above) or via [Service Discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config), especially given that [Cephadm is moving to SD](https://github.com/ceph/ceph/pull/46400/) for Node-exporter, HAProxy and mgr/Prometheus metrics.

To me that's another story. First, of all enabling this will break users relaying on the package that do not run cephadm. Second, even if we only take into account cephadm, this could be enabled in the future after this PR. For instance, you might flip the multi cluster boolean to true by default and regenerate the files or even introduce some other packages with the dashboards and alerts.

Reply:
Hi @epuertat, FYI @bosc0 is leaving CERN before the end of the month and I am leaving early next month. So if we can rule this PR out before it would be pretty nice :D.

Reply:
😲 sorry to hear that, guys (I mean it'll be a loss for the Ceph community too).

Ok, I'm not closely following the Prometheus mixins trends, so you'll know this much better than I do.

Rather than embedding the cluster ID in the alert text message, why not using a new custom annotation or label (e.g.: `ceph_cluster_fsid`) for that? That would allow for easy filtering/grouping. What do you think about it?

Reply:
> 😲 sorry to hear that, guys (I mean it'll be a loss for the Ceph community too).

Thanks! Actually I will join another company that use Ceph heavily but I won't be working on Ceph myself though... Maybe I would still contribute from time to time though... 

> Ok, I'm not closely following the Prometheus mixins trends, so you'll know this much better than I do.
> 
> Rather than embedding the cluster ID in the alert text message, why not using a new custom annotation or label (e.g.: `ceph_cluster_fsid`) for that? That would allow for easy filtering/grouping. What do you think about it?

Well this is already possible https://github.com/ceph/ceph/blob/1c260e88f28864d7e1125161ef9099a41fc3e12e/monitoring/ceph-mixin/config.libsonnet#L5-L6. For now it is configured to use the label "cluster" by default which would be in our non-cephadm use case the name of the cluster that we give to each of them. So If you add cephadm integration later on you would easily be able to change the default label and figure out if you want to have a granularity by fsid, by a name provided by the user somehow or even autogenerated or whatever you think is appropriate....

BTW we are discussing this for alert but as this is already implemented for the dashboards